src/exo2 - Combinators/exo2.effect.solution.ts

```diff
@@ -0,0 +1,198 @@
+//                                   SETUP                                   //
+///////////////////////////////////////////////////////////////////////////////
+
+// We are developing a small game, and the player can control either one of
+// three types of characters, mainly differentiated by the type of damage they
+// can put out.
+
+// Our main `Character` type is a simple union of all the concrete character
+// types.
+export type Character = Warrior | Wizard | Archer;
+
+// We have three types of `Damage`, each corresponding to a character type.
+export enum Damage {
+  Physical = 'Physical damage',
+  Magical = 'Magical damage',
+  Ranged = 'Ranged damage',
+}
+
+// A `Warrior` only can output physical damage.
+export class Warrior {
+  smash() {
+    return Damage.Physical;
+  }
+
+  toString() {
+    return 'Warrior';
+  }
+}
+
+// A `Wizard` only can output magical damage.
+export class Wizard {
+  burn() {
+    return Damage.Magical;
+  }
+
+  toString() {
+    return 'Wizard';
+  }
+}
+
+// An `Archer` only can output ranged damage.
+export class Archer {
+  shoot() {
+    return Damage.Ranged;
+  }
+
+  toString() {
+    return 'Archer';
+  }
+}
+
+// We also have convenient type guards to help us differentiate between
+// character types when given a `Character`.
+
+export const isWarrior = (character: Character): character is Warrior => {
+  return (character as Warrior).smash !== undefined;
+};
+
+export const isWizard = (character: Character): character is Wizard => {
+  return (character as Wizard).burn !== undefined;
+};
+
+export const isArcher = (character: Character): character is Archer => {
+  return (character as Archer).shoot !== undefined;
+};
+
+// Finally, we have convenient and expressive error types, defining what can
+// go wrong in our game:
+// - the player can try to perform an action without first choosing a character as the attacker
+// - the player can try to perform the wrong action for a character class
+
+export enum Exo2FailureType {
+  NoAttacker = 'Exo2FailureType_NoAttacker',
+  InvalidAttacker = 'Exo2FailureType_InvalidAttacker',
+}
+
+export type NoAttackerFailure = Failure<Exo2FailureType.NoAttacker>;
+export const noAttackerFailure = Failure.builder(Exo2FailureType.NoAttacker);
+
+export type InvalidAttackerFailure = Failure<Exo2FailureType.InvalidAttacker>;
+export const invalidAttackerFailure = Failure.builder(
+  Exo2FailureType.InvalidAttacker,
+);
+
+///////////////////////////////////////////////////////////////////////////////
+//                                  EITHER                                   //
+///////////////////////////////////////////////////////////////////////////////
+
+// Helper functions for checking character types
+// Note: In Effect, Either is Either<Right, Left> (opposite to fp-ts)
+export const checkAttackerAndSmash = (attacker: Option.Option<Character>): Either.Either<Damage, NoAttackerFailure | InvalidAttackerFailure> => {
```

Comment:
This exercise solution is not matching at all the one from fp-ts -> I recommend doing it manually here and reproduce more what was done on the fp-ts side and simply adapt the import (and little things like fromPredicate -> liftPredicate)